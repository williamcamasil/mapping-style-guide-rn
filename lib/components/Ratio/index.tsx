import React, { useCallback, useState, useMemo } from 'react';
import {
  FlexStyle,
  ImageLoadEventData,
  ImageStyle,
  LayoutChangeEvent, NativeSyntheticEvent, StyleProp, StyleSheet, TextStyle, View, ViewProps, ViewStyle,
} from 'react-native';

import { useViewStyles } from '../../hooks';
import {
  calculateRatioBy, ChildrenRenderFuncType, isRenderFunc, RatioSizeType, RatioSourceType,
} from './utils';

export type { RatioSizeType } from './utils';

type RatioPropsType<InnerStyle> = Omit<ViewProps, 'children'> & {
  children: React.ReactNode | ChildrenRenderFuncType<InnerStyle>;
  originalWidth?: number;
  originalHeight?: number;
  ratioSource?: RatioSourceType;
  innerStyle?: StyleProp<InnerStyle>;
};

function Ratio<InnerStyle extends (FlexStyle | ViewStyle | TextStyle | ImageStyle) = RatioSizeType>({
  children,
  style,
  onLayout,
  originalWidth,
  originalHeight,
  ratioSource,
  innerStyle,
  ...others
}: RatioPropsType<InnerStyle>) {
  const [originalSize, setOriginalSize] = useState<RatioSizeType>({
    width: originalWidth ?? 0,
    height: originalHeight ?? 0,
  });
  const [containerSize, setContainerSize] = useState<RatioSizeType>();

  const handleLayout = useCallback((event: LayoutChangeEvent) => {
    onLayout?.(event);
    const { width, height } = event.nativeEvent.layout;
    setContainerSize({
      width, height,
    });
  }, [onLayout]);

  const onImageLoadCallback = useCallback((event: NativeSyntheticEvent<ImageLoadEventData>) => {
    const { height, width } = event.nativeEvent.source;
    setOriginalSize({
      height,
      width,
    });
  }, []);

  const ratioSize = useMemo(
    () => calculateRatioBy(ratioSource, originalSize, containerSize),
    [containerSize, originalSize, ratioSource],
  );

  const ratioStyles = useViewStyles(() => [
    ratioSize,
    style,
  ], [ratioSize, style]);

  const ratioStyle = useMemo(() => ({
    ...StyleSheet.flatten(innerStyle),
    ...ratioSize,
  }), [innerStyle, ratioSize]);

  const renderChildren = () => {
    if (isRenderFunc(children)) {
      return children(
        ratioStyle,
        onImageLoadCallback,
      );
    }

    return children;
  };

  return (
    <View onLayout={handleLayout}>
      <View {...others} style={ratioStyles}>
        {renderChildren()}
      </View>
    </View>
  );
}

Ratio.defaultProps = {
  ratioSource: 'width',
};

export default Ratio;
