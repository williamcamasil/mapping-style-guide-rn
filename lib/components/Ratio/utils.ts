import { ImageLoadEventData, NativeSyntheticEvent } from 'react-native';

export type RatioSourceType = 'width' | 'height';

export type RatioSizeType = {
  width: number;
  height: number;
};

export type ChildrenRenderFuncType<InnerStyle> = (
  style: (InnerStyle extends (infer U)[] ? U : InnerStyle) & {
    width: number | string;
    height: number | string;
  },
  onImageLoad: (event: NativeSyntheticEvent<ImageLoadEventData>) => void | undefined,
) => JSX.Element;

const calculateByHeight = (originalSize: RatioSizeType, containerSize: RatioSizeType) => ({
  width: (containerSize.height * originalSize.width) / originalSize.height,
  height: containerSize.height,
});

const calculateByWidth = (originalSize: RatioSizeType, containerSize: RatioSizeType) => ({
  width: containerSize.width,
  height: (containerSize.width * originalSize.height) / originalSize.width,
});

export const calculateRatioBy = (
  ratioSource?: RatioSourceType,
  originalSize?: RatioSizeType,
  containerSize?: RatioSizeType,
) => {
  if (!containerSize || !originalSize?.width || !originalSize?.height) {
    return {
      height: '100%',
      width: '100%',
    };
  }

  if (ratioSource === 'height') {
    return calculateByHeight(originalSize, containerSize);
  }

  return calculateByWidth(originalSize, containerSize);
};

export function isRenderFunc<InnerStyle>(
  children: React.ReactNode | ChildrenRenderFuncType<InnerStyle>,
): children is ChildrenRenderFuncType<InnerStyle> {
  return typeof children === 'function';
}
