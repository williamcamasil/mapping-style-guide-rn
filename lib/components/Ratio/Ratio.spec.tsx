import * as React from 'react';
import { Image, StyleSheet, View } from 'react-native';

import { render, fireEvent } from '@testing-library/react-native';

import Ratio, { RatioSizeType } from '.';
import ImageExample from './image-example.png';

const fakeLayoutEvent = {
  nativeEvent: {
    layout: {
      width: 320,
      height: 1080,
    },
  },
};

const fakeImageLoadEvent = {
  nativeEvent: {
    source: {
      width: 320,
      height: 1080,
    },
  },
};

describe('Ratio snapshot', () => {
  it('by width', () => {
    const result = render(
      <Ratio
        testID="ratio-instance"
        originalWidth={1201}
        originalHeight={601}
        ratioSource="width"
      >
        {size => (
          <Image
            testID="internal-image"
            source={ImageExample}
            style={size}
          />
        )}
      </Ratio>,
    );

    fireEvent(result.getByTestId('ratio-instance'), 'layout', fakeLayoutEvent);

    expect(result.toJSON()).toMatchSnapshot();
  });

  it('by height', () => {
    const result = render(
      <Ratio
        testID="ratio-instance"
        originalWidth={1201}
        originalHeight={601}
        ratioSource="height"
      >
        {(size: RatioSizeType) => (
          <Image
            testID="internal-image"
            source={ImageExample}
            style={size}
          />
        )}
      </Ratio>,
    );

    fireEvent(result.getByTestId('ratio-instance'), 'layout', fakeLayoutEvent);

    expect(result.toJSON()).toMatchSnapshot();
  });

  it('children element', () => {
    const result = render(
      <Ratio
        testID="ratio-instance"
        originalWidth={1201}
        originalHeight={601}
      >
        <View
          testID="internal-view"
          style={{
            width: '100%',
            height: '100%',
          }}
        />
      </Ratio>,
    );

    fireEvent(result.getByTestId('ratio-instance'), 'layout', fakeLayoutEvent);

    expect(result.toJSON()).toMatchSnapshot();
  });
});

describe('Ratio styles', () => {
  it('should calculate size by width', () => {
    const result = render(
      <Ratio
        testID="ratio-instance"
        originalWidth={1201}
        originalHeight={601}
        ratioSource="width"
      >
        {(size: RatioSizeType) => (
          <Image
            testID="internal-image"
            source={ImageExample}
            style={size}
          />
        )}
      </Ratio>,
    );

    fireEvent(result.getByTestId('ratio-instance'), 'layout', fakeLayoutEvent);

    const imageInstance = result.getByTestId('internal-image');
    const imageStyles = StyleSheet.flatten(imageInstance.props.style);

    expect(imageStyles).toHaveProperty('width', 320);
    expect(imageStyles).toHaveProperty('height', 160.1332223147377);
  });

  it('should calculate size by height', () => {
    const result = render(
      <Ratio
        testID="ratio-instance"
        originalWidth={1201}
        originalHeight={601}
        ratioSource="height"
      >
        {(size: RatioSizeType) => (
          <Image
            testID="internal-image"
            source={ImageExample}
            style={size}
          />
        )}
      </Ratio>,
    );

    fireEvent(result.getByTestId('ratio-instance'), 'layout', fakeLayoutEvent);

    const imageInstance = result.getByTestId('internal-image');
    const imageStyles = StyleSheet.flatten(imageInstance.props.style);

    expect(imageStyles).toHaveProperty('width', 2158.2029950083192);
    expect(imageStyles).toHaveProperty('height', 1080);
  });

  it('should pass inner styles', () => {
    const result = render(
      <Ratio
        testID="ratio-instance"
        originalWidth={1201}
        originalHeight={601}
        innerStyle={{
          borderRadius: 24,
          overflow: 'hidden',
        }}
      >
        {size => (
          <Image
            testID="internal-image"
            source={ImageExample}
            style={size}
          />
        )}
      </Ratio>,
    );

    fireEvent(result.getByTestId('ratio-instance'), 'layout', fakeLayoutEvent);

    const imageInstance = result.getByTestId('internal-image');
    const imageStyles = StyleSheet.flatten(imageInstance.props.style);

    expect(imageStyles).toHaveProperty('borderRadius', 24);
    expect(imageStyles).toHaveProperty('overflow', 'hidden');
  });

  it('should load size from remote image inner styles', () => {
    const result = render(
      <Ratio
        innerStyle={{
          borderRadius: 24,
          overflow: 'hidden',
        }}
      >
        {(innerStyle, onImageLoadCallback) => (
          <Image
            testID="internal-image"
            source={{
              uri: 'https://upload.wikimedia.org/wikipedia/commons/thumb/3/3f/Placeholder_view_vector.svg/681px-Placeholder_view_vector.svg.png',
            }}
            onLoad={onImageLoadCallback}
            style={innerStyle}
          />
        )}
      </Ratio>,
    );

    fireEvent(result.getByTestId('internal-image'), 'load', fakeImageLoadEvent);

    const imageInstance = result.getByTestId('internal-image');
    const imageStyles = StyleSheet.flatten(imageInstance.props.style);

    expect(imageStyles).toHaveProperty('borderRadius', 24);
    expect(imageStyles).toHaveProperty('overflow', 'hidden');
  });
});
