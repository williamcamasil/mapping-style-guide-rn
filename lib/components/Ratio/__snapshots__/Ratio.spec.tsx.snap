// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ratio snapshot by height 1`] = `
<View
  onLayout={[Function]}
>
  <View
    style={
      {
        "height": 1080,
        "width": 2158.2029950083192,
      }
    }
    testID="ratio-instance"
  >
    <Image
      source={
        {
          "testUri": "../../../lib/components/Ratio/image-example.png",
        }
      }
      style={
        {
          "height": 1080,
          "width": 2158.2029950083192,
        }
      }
      testID="internal-image"
    />
  </View>
</View>
`;

exports[`Ratio snapshot by width 1`] = `
<View
  onLayout={[Function]}
>
  <View
    style={
      {
        "height": 160.1332223147377,
        "width": 320,
      }
    }
    testID="ratio-instance"
  >
    <Image
      source={
        {
          "testUri": "../../../lib/components/Ratio/image-example.png",
        }
      }
      style={
        {
          "height": 160.1332223147377,
          "width": 320,
        }
      }
      testID="internal-image"
    />
  </View>
</View>
`;

exports[`Ratio snapshot children element 1`] = `
<View
  onLayout={[Function]}
>
  <View
    style={
      {
        "height": 160.1332223147377,
        "width": 320,
      }
    }
    testID="ratio-instance"
  >
    <View
      style={
        {
          "height": "100%",
          "width": "100%",
        }
      }
      testID="internal-view"
    />
  </View>
</View>
`;
